### PR TITLE
fix: Prevent double-linkification of mentions and markdown links

### PR DESCRIPTION
**Pull request by menobass**
## Summary

- **Mentions showing as raw markdown** (`[@bitterirony](profile://bitterirony)` visible as literal text): `preprocessForMarkdown` had its own `@username` regex that ran *after* `linkifyMentions` had already formed the link, corrupting it into `[[**@user**](profile://user)](profile://user)` — invalid nested syntax the renderer couldn't parse.
- **User-written `[text](url)` links being mangled**: `linkifyUrls` had a broken guard that tested the URL string itself rather than the surrounding context, so `[My blog](https://url.com)` was double-wrapped into `[My blog]([https://url.com](https://url.com))`.
- **Third bug found by tests**: The `isInsideMarkdownLink` guard in `linkifyMentions` failed when `[@user]` appeared at the start of a string — the `[` was captured as `pre`, absent from `beforeMatch`, so the bracket-count check returned false.

## Changes

- `utils/htmlPreprocessing.ts` — removed the `@username` replacement (mention linkification belongs to `linkifyMentions`, not here)
- `utils/linkifyMentions.ts` — extracted from `Snap.tsx` into a shared util; fixed the start-of-string guard
- `utils/linkifyUrls.ts` — extracted from `Snap.tsx` into a shared util; fixed the context guard to check `charBefore`/`charAfter` in the original string
- `app/components/Snap.tsx` — imports from the new util files
- `app/components/PostBody.tsx` — now calls `linkifyMentions` on its markdown fallback path
- `tests/linkification.test.ts` — 23 tests covering the full pipeline

## Test plan

- [ ] All 23 unit tests pass (`npx jest tests/linkification.test.ts`)
- [ ] Snap with `@mention` renders as a clickable blue link, not raw text
- [ ] Snap with `[@user](profile://user)` stored on-chain renders correctly (not double-wrapped)
- [ ] `[My blog](https://url.com)` written by a user renders as a clickable link
- [ ] Bare `https://url.com` in a snap still gets auto-linked
- [ ] YouTube/Hive post URLs are still left unwrapped for their dedicated renderers

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved linkification: bare @mentions and plain URLs are consistently turned into clickable links before Markdown rendering; avoids double-linking, false matches (emails, excluded URL types), and linkification inside HTML attributes. Mentions normalize case and length rules apply; long URLs show truncated display text.

* **New Features**
  * Centralized/shared linkification utilities to standardize behavior across the app.

* **Tests**
  * Added comprehensive tests covering mention/URL linkification and edge cases.

### CODE REVIEW

File: tests/linkification.test.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    // This was the bug: [My blog](https://puppy.com) was being mangled into
+    // [My blog]([https://puppy.com](https://puppy.com))
+    const input = '[My favorite puppy blog](https://puppy.com)';
+    expect(linkifyUrls(input)).toBe(input);
+  });
+
+  it('does NOT double-wrap a URL already inside a markdown link mid-sentence', () => {
+    const input = 'Check out [this site](https://example.com) for details.';
+    expect(linkifyUrls(input)).toBe(input);
+  });
```
[the remark this answers is not in the document]
Comment (menobass) [reply tests/linkification.test.ts:110]: Added both cases in `1e17da9`: `[text](url "title")` and `[https://example.com](https://example.com)` (URL as link text). Both pass — the updated guards in `linkifyUrls` handle them correctly.

File: utils/linkifyMentions.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * Skips mentions that are already inside a markdown link [text](url) to
+ * prevent double-linkification from corrupting the syntax.
+ *
+ * Produces:  [@username](profile://username)
+ * The visual bold/colour treatment is handled downstream via linkStyles.mention.
+ */
+export function linkifyMentions(text: string): string {
+  return text.replace(
+    /(^|[^\w/@])@([a-z0-9\-\.]{3,16})(?![a-z0-9\-\.])/gi,
```
[the remark this answers is not in the document]
Comment (menobass) [reply utils/linkifyMentions.ts]: Fixed in `1e17da9`. Updated the capture group to `[a-z][a-z0-9\-.]{1,14}[a-z0-9]` — must start with a letter, end with alphanumeric, 3–16 chars total. Also added `.toLowerCase()` normalisation since Hive usernames are always lowercase on-chain. Tests added for `@-alice` and `@alice-` (both rejected) and uppercase input normalisation.

File: utils/linkifyUrls.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+export function linkifyUrls(text: string): string {
+  return text.replace(
+    /(https?:\/\/[\w.-]+(?:\/[\w\-./?%&=+#@]*)?)/gi,
+    (url, _p1, offset: number, string: string) => {
+      // A URL is already inside a markdown link [text](url) when the character
+      // immediately before it is '(' and immediately after it is ')'.
+      const charBefore = string[offset - 1];
+      const charAfter = string[offset + url.length];
+      if (charBefore === '(' && charAfter === ')') return url;
+
```
[the remark this answers is not in the document]
Comment (menobass) [reply utils/linkifyUrls.ts]: Fixed in `1e17da9`. Replaced the narrow `charAfter === ')'` check with two proper context checks:
1. **URL as link text** (`[https://x](https://x)`): bracket counting — if there are more open brackets than close brackets before the URL and `](` follows, skip it.
2. **URL as link destination with optional title** (`[text](https://x "title")`): `charBefore === '('` combined with a regex that allows optional whitespace + title string before the closing `)`.

Both new edge cases are covered by the added tests.

File: utils/linkifyUrls.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      // Also skip HTML href attributes
+      const nearby = string.substring(Math.max(0, offset - 8), offset);
+      if (/href=["']?$/.test(nearby)) return url;
```
[the remark this answers is not in the document]
Comment (menobass) [reply utils/linkifyUrls.ts]: Fixed in `1e17da9`. Switched to `\bhref\s*=\s*["']?$/i` and extended the lookback window to 24 characters to accommodate the extra whitespace. A test for `HREF="..."` is included in the suite.

File: utils/linkifyMentions.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * The visual bold/colour treatment is handled downstream via linkStyles.mention.
+ */
+export function linkifyMentions(text: string): string {
+  return text.replace(
+    // Hive usernames: start with a letter, end with alphanumeric, 3–16 chars total.
+    // Middle chars may include letters, digits, hyphens, and dots.
+    /(^|[^\w/@])@([a-z][a-z0-9\-.]{1,14}[a-z0-9])(?![a-z0-9\-.])/gi,
+    (match, pre, username, offset, string) => {
+      const beforeMatch = string.substring(0, offset);
+      const afterMatch = string.substring(offset + match.length);
```
[the remark this answers is not in the document]
Comment (menobass) [reply utils/linkifyMentions.ts]: Fixed in `56b9bd7`. Added explicit types to all callback params (`match: string, pre: string, username: string, offset: number, source: string): string`) and renamed `string` → `source` to avoid shadowing the built-in.

File: utils/linkifyUrls.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ * HTML href attribute, to prevent double-wrapping which produces invalid nested
+ * link syntax and causes the renderer to show raw text.
+ *
+ * Non-normal URLs (Hive posts, embedded media, YouTube, etc.) are returned
+ * unchanged so their dedicated renderers can handle them.
+ */
+export function linkifyUrls(text: string): string {
+  return text.replace(
+    /(https?:\/\/[\w.-]+(?:\/[\w\-./?%&=+#@]*)?)/gi,
+    (url, _p1, offset: number, string: string) => {
```
[the remark this answers is not in the document]
Comment (menobass) [reply utils/linkifyUrls.ts]: Fixed in `56b9bd7`. Removed the capture group (`/(url)/` → `/url/`), typed the callback as `(url: string, offset: number, source: string): string`, and renamed `string` → `source` throughout.

File: utils/linkifyUrls.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      const openBrackets = (beforeMatch.match(/\[/g) || []).length;
+      const closeBrackets = (beforeMatch.match(/\]/g) || []).length;
+      if (openBrackets > closeBrackets && afterMatch.includes('](')) return url;
+
+      // Skip if the URL is the href of a markdown link, including optional
+      // title: [text](https://... "optional title") or [text](https://...)
+      if (
+        charBefore === '(' &&
+        /^\s*(?:"[^"]*"|'[^']*'|\([^)]*\))?\)/.test(afterMatch)
+      ) {
```
Comment (menobass) [inline utils/linkifyUrls.ts:34]: Fixed in `56b9bd7`. Replaced `charBefore === '('` with `/\]\(\s*$/.test(beforeMatch)` — the `(` must be immediately preceded by `]` to qualify as a markdown link destination. Plain parenthesised URLs like `(https://example.com)` in prose are now linkified correctly. Test added to cover this case (30 total, all passing).

File: tests/linkification.test.ts [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+  it('normalises uppercase mentions to lowercase', () => {
+    expect(linkifyMentions('Hello @Alice!')).toBe(
+      'Hello [@alice](profile://alice)!'
+    );
+  });
+
+  it('does NOT wrap a URL inside an HTML href with uppercase HREF', () => {
+    const input = '<a HREF="https://example.com">click</a>';
+    expect(linkifyUrls(input)).toBe(input);
+  });
```
[the remark this answers is not in the document]
Comment (menobass) [reply tests/linkification.test.ts]: Fixed in `18aa79e` — removed from the `linkifyMentions` block and placed it in the `linkifyUrls` block alongside the other href attribute tests.

File: tests/linkification.test.ts [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+  it('does NOT wrap a URL inside an HTML href attribute', () => {
+    const input = '<a href="https://example.com">click</a>';
+    expect(linkifyUrls(input)).toBe(input);
+  });
+
+  it('does NOT wrap a URL inside an HTML href attribute without quotes', () => {
+    const input = "<a href='https://example.com'>click</a>";
+    expect(linkifyUrls(input)).toBe(input);
```
[the remark this answers is not in the document]
Comment (menobass) [reply tests/linkification.test.ts]: Fixed in `18aa79e` — renamed to "with single quotes".

File: utils/linkifyUrls.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+      const closeBrackets = (beforeMatch.match(/\]/g) || []).length;
+      if (openBrackets > closeBrackets && afterMatch.includes('](')) return url;
+
+      // Skip if the URL is the href of a markdown link [text](url "title").
+      // Require beforeMatch to end with "](" so plain parenthesised URLs like
+      // (https://example.com) are still linkified.
+      if (
+        /\]\(\s*$/.test(beforeMatch) &&
+        /^\s*(?:"[^"]*"|'[^']*'|\([^)]*\))?\)/.test(afterMatch)
+      ) {
```
[the remark this answers is not in the document]
Comment (menobass) [reply utils/linkifyUrls.ts:34]: Fixed in `18aa79e`. Updated `beforeMatch` check to `/\]\(\s*<?\s*$/` (allows optional `<`) and `afterMatch` check to `/^\s*>?\s*(?:(?:"[^"]*"|'[^']*'|\([^)]*\))\s*)?\)/` (allows optional `>`). Test added for `[label](<https://example.com>)` — 31 tests, all passing.

**Comment by menobass** [comment]
@coderabbit review

File: utils/linkifyUrls.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ * HTML href attribute, to prevent double-wrapping which produces invalid nested
+ * link syntax and causes the renderer to show raw text.
+ *
+ * Non-normal URLs (Hive posts, embedded media, YouTube, etc.) are returned
+ * unchanged so their dedicated renderers can handle them.
+ */
+export function linkifyUrls(text: string): string {
+  return text.replace(
+    /https?:\/\/[\w.-]+(?:\/[\w\-./?%&=+#@]*)?/gi,
+    (url: string, offset: number, source: string): string => {
```
[the remark this answers is not in the document]
Comment (menobass) [reply utils/linkifyUrls.ts]: Fixed in `1fe54be`. Updated the regex from `/https?:\/\/[\w.-]+(?:\/[\w\-./?%&=+#@]*)?/gi` to `/https?:\/\/[\w.-]+(?::\d+)?(?:[/?#][\w\-./?%&=+#@:;]*)?/gi`:
- `(?::\d+)?` — optional port
- `(?:[/?#]...)` instead of `(?:\/...)` — allows query (`?`) or fragment (`#`) to start the optional tail without requiring a path segment first

Three regression tests added and passing (34 total).

**Review by eddiespino — APPROVED** [review]

**Review by eddiespino — APPROVED** [review]

**Comment by eddiespino** [comment]
I can't see Homer Simpson wife.
